### DISCUSSION TITLE
Route app groups

### DISCUSSION THREAD

**Discussion by jamesrweb** [clue/framework-x#216] [Ideas]
It would be nice if we could implement route groups, for example:

```php
$app->group('/api', function (AppGroup $apiGroup) {
    $apiGroup->group('/v1', function (AppGroup $v1Group) {
        $v1Group->group('/books', function (AppGroup $booksGroup) {
          $booksGroup->get('/:id', FindBookController::class);
          $booksGroup->post('/', CreateBookController::class);
          $booksGroup->put('/:id', UpdateBookController::class);
          $booksGroup->delete('/:id', DeleteBookController::class);
        });
        $v1Group->group('/users', function (AppGroup $usersGroup) {
          $usersGroup->get('/:id', FindUserController::class);
          $usersGroup->post('/', CreateUserController::class);
          $usersGroup->put('/:id', UpdateUserController::class);
          $usersGroup->delete('/:id', DeleteUserController::class);
        });
    });
});
```

We could also pass such `AppGroup` arguments to an invokable class provided to the `->group()` method also, for example:

```php
class BooksGroup implements InvokableAppGroup {
    public function __invoke(AppGroup $group) {
        $group->get('/:id', FindBookController::class);
        $group->post('/', CreateBookController::class);
        $group->put('/:id', UpdateBookController::class);
        $group->delete('/:id', DeleteBookController::class);
    }
}

class UsersGroup implements InvokableAppGroup {
    public function __invoke(AppGroup $group) {
        $group->get('/:id', FindUserController::class);
        $group->post('/', CreateUserController::class);
        $group->put('/:id', UpdateUserController::class);
        $group->delete('/:id', DeleteUserController::class);
    }
}

$app->group('/api', function (AppGroup $apiGroup) {
    $apiGroup->group('/v1', function (AppGroup $v1Group) {
        $v1Group->group('/books', BooksGroup::class);
        $v1Group->group('/users', UsersGroup::class);
    });
});
```

This way we could create clean groupings in a familiar way to how Controllers currently work.

I would personally find this extremely helpful at decluttering my `$app->*` calls. Furthermore, you may have noticed that I took the liberty of adding a `InvokableAppGroup` interface above, this is something we are missing for Controllers, etc currently but could be a really nice addition to the framework so that such classes can be implemented via contract rather than assumption. Of course this is a separate topic but it would be nice to have such interfaces for core concepts as native parts of the framework in future.

**Comment by SimonFrings** [comment]
Hey @jamesrweb, thanks for bringing in your idea and showing us through examples what this might look like :+1:

I agree, this is a wanted feature, as also asked for in https://github.com/clue/framework-x/discussions/134. We want to see this being part of Framework X as well, so this is definitely on our list. We're currently working on the next major version for ReactPHP (v3) which takes up a lot of our time and will not only improve ReactPHP itself, it also enhances Framework X and other projects build on top of ReactPHP.

**Reply by jamesrweb** [reply]
Any rough idea when v3 will be done and framework-x will get attention again?


Some side topics:
- The video on the homepage doesn't play
- The documentation could be better rounded with more examples of not just simple requests but streaming requests and responses or long lived components generally
- Another thing which would be a godsend is catch all routes such as: `$app->get("*", Handler::class);`

Any thoughts on these?

**Reply by SimonFrings** [reply]
@jamesrweb We will work on ReactPHP v3 for the next couple of months, but I can't tell you an exact date when this will be released. At the same time, we want to continuously work on Framework X, @clue and I also talked about our next steps for the project a couple of days ago.

> The video on the homepage doesn't play

Yes, we're aware of that. The current implementation acts as a placeholder for a future video, I've opened a ticket regarding this topic here: https://github.com/clue/framework-x-website/issues/33

> The documentation could be better rounded with more examples of not just simple requests but streaming requests and responses or long lived components generally

I totally agree with you and this perfectly aligns with our latest discussion in our team. We plan on improving the current "under construction" parts of our documentation, working towards a stable release for Framework X. Any specific ideas, plans or contributions for our documentation are very welcome, just open a new discussion or add ideas to a already existing one regarding the same topic and we're happy to take a look!

> Another thing which would be a godsend is catch all routes such as: $app->get("*", Handler::class);

This is actually already possible, but it needs some better documentation. You can take a look at the [`examples/index.php`](https://github.com/clue/framework-x/blob/main/examples/index.php) and you'll find a few examples like this (line 108):
```php
$app->get('/source/{path:.*}', new FrameworkX\FilesystemHandler(dirname(__DIR__)));
```
I think it is worth to open up a new discussion for this, might be a first step to give this more visibility :+1:

**Reply by jamesrweb** [reply]
@SimonFrings thanks for the detailed answer! So far, so good with the planning for the next stages then I would say!

**Comment by SimonFrings** [comment]
To keep you updated, we released the first version of Framework X yesterday ([v0.16.0](https://github.com/clue/framework-x/releases/tag/v0.16.0)), along with all the other prior tagged versions! 🎉

With these new versions out, we can now create smooth upgrade paths between versions and it also makes it easier for us to introduce BC breaks in the future. While working on the new version, we also made a list of all the features and prototypes that we have started working on in the past, but never finished due to the release situation (the situation being that we had none 😅).

Adding route groups is one of these topics, and we're happy to further look into this while working on the next versions for Framework X (don't know which version specifically yet). I also updated https://github.com/clue/framework-x/pull/241 recently which was a pull request to introduce route groups to Framework X. The PR was a good starting point for this topic, but we'll probably make changes to the `RouterHandler` and the overall handler structure in advance. Once this is done we'll look further into supporting route groups.

I'll keep this ticket here updated as we make progress :+1:

If you need this (more) urgently, you can also reach out to us or help us prioritizing by [becoming as sponsor](https://github.com/sponsors/reactphp) ❤️
